contoso_manufacturing/developer/webapp-decode/yolov8.py

```diff
@@ -22,37 +24,64 @@ def __init__(self, rtsp_url, class_names, input_shape, color_palette, confidence
-
+        self.grpc_client = make_grpc_client(ovms_url)
+        self.stopped = False
+        self.lock = threading.Lock()
+        # captures the frames and preprocesses them
+        self.preprocessed_frames_queue = queue.Queue(maxsize=150)
+        # captures the frames and outputs from the inference
+        self.inferenced_frames_queue = queue.Queue(maxsize=150)
+        # postprocesses the frames
+        self.postprocessed_frames_queue = queue.Queue(maxsize=150)
+        self.capture_thread = threading.Thread(target=self.capture_frames)
+        self.capture_thread.start()
+        self.postprocess_thread = threading.Thread(target=self.postprocess_frames)
+        self.postprocess_thread.start()
+        self.inference_thread = threading.Thread(target=self.run_inference)
+        self.inference_thread.start()
         # Track frames and inference processing time for displaying FPS performance metrics 
         self.total_inference_time = 0.0
         self.inference_fps = 0.0
         self.total_fps = 0.0
         self.total_frames = 0
         self.start_time = time.time()
 
-        self.cap = cv2.VideoCapture(rtsp_url)
-        self.grpc_client = make_grpc_client(ovms_url)
-        
-        if not self.cap.isOpened():
-            print("Error: Unable to open video source.")
-        else:
-            # Lee un frame para determinar el tamaño de los frames del video
-            ret, frame = self.cap.read()
-            if ret:
-                self.img_height, self.img_width = frame.shape[:2]
-                print(f"Image dimensions: {self.img_width}x{self.img_height}")
-            else:
-                print("Failed to grab frame to set image dimensions")
+    def capture_frames(self):
+        cap = cv2.VideoCapture(self.rtsp_url)
+        while not self.stopped:
+            ret, frame = cap.read()
+            if not ret:
+                print("Failed to grab frame")
+                break
+            
+            preprocessed_frame = self.preprocess(frame)
+            frame_tuple = (frame, preprocessed_frame)
+            while not self.stopped and self.preprocessed_frames_queue.full():
+                time.sleep(0.005)
+            self.log("Adding frame to preprocessed frames queue...")
+            self.preprocessed_frames_queue.put(frame_tuple)
+        cap.release()
+
+    def postprocess_frames(self):
+        while not self.stopped:
+            if(self.inferenced_frames_queue.empty()):
+                time.sleep(0.005)
+                continue
+            self.log("Postprocessing frames...")
+            frame, outputs = self.inferenced_frames_queue.get()
+            
+            postprocessed_frame = self.postprocess(frame, outputs)
+            while not self.stopped and self.postprocessed_frames_queue.full():
+                time.sleep(0.005)
+            self.log("Adding postprocessed frame to postprocessed frames queue...")
+
+            self.postprocessed_frames_queue.put(postprocessed_frame)        
```

Comment:
I think queue.put() is a blocking method (unless you specify blocking=False) so this could hang during shutdown: once self.stopped is True, you will drop out of the sleep loop even if 'queue.full' is still true.  Possible mitigations:

1. Add the test for self.postprocessed_frames_queue.full() at the beginning of the method (at the same time you test self.inferenced_frames_queue.empty()).  That guarantees the entire method can always complete without blocking.
2. Add "blocking=false" to the put() call and catch (and ignore) the queue.Full exception which might happen when the thread is shutting down.
3. Do nothing since this is just a sample app. and we can just live with it.

Options 1 and 2 are both pretty straightforward changes but I'll let you decide if it's worth the trouble to rebuild / retest in this instance (especially since we are probably going to move this code over to a different repo anyway)..

Reply:
Added explicit check for queue is full or empty before calling the put/get calls so as to not block the thread.